File: frontend/src/services/contratoService.js
import api from "./api";

const API_URL = "/contratos";

const listar = async () => {
  const response = await api.get(API_URL);
  return response.data;
};

const buscarPorId = async (id) => {
  const response = await api.get(`${API_URL}/${id}`);
  return response.data;
};

const crear = async (contrato) => {
  const response = await api.post(API_URL, contrato);
  return response.data;
};

const actualizar = async (id, contrato) => {
  const response = await api.put(`${API_URL}/${id}`, contrato);
  return response.data;
};

const eliminar = async (id) => {
  const response = await api.delete(`${API_URL}/${id}`);
  return response.data;
};

const listarPorEstado = async (estado) => {
  const response = await api.get(`${API_URL}/estado/${estado}`);
  return response.data;
};

const listarPorCliente = async (idCliente) => {
  const response = await api.get(`${API_URL}/cliente/${idCliente}`);
  return response.data;
};

const listarProximosAVencer = async (dias = 30) => {
  const response = await api.get(`${API_URL}/proximos-vencer`, {
    params: { dias },
  });
  return response.data;
};

const listarVencidos = async () => {
  const response = await api.get(`${API_URL}/vencidos`);
  return response.data;
};

const cambiarEstado = async (id, estado) => {
  const response = await api.patch(`${API_URL}/${id}/estado`, {}, {
    params: { estado },
  });
  return response.data;
};

const contratoService = {
  listar,
  buscarPorId,
  crear,
  actualizar,
  eliminar,
  listarPorEstado,
  listarPorCliente,
  listarProximosAVencer,
  listarVencidos,
  cambiarEstado,
};

export default contratoService;

File: backend/src/main/java/co/edu/uniquindio/backend/service/ContratoService.java
package co.edu.uniquindio.backend.service;

import co.edu.uniquindio.backend.estructuras.listas.SimpleList.LinkedSimpleList;
import co.edu.uniquindio.backend.model.Contrato;
import co.edu.uniquindio.backend.repository.ContratoRepository;
import org.springframework.stereotype.Service;

@Service
public class ContratoService {

    private final ContratoRepository contratoRepository;

    public ContratoService(ContratoRepository contratoRepository) {
        this.contratoRepository = contratoRepository;
    }

    public Contrato[] listar() {
        actualizarEstadosAutomaticos();
        return convertirAArreglo(contratoRepository.listar());
    }

    public Contrato buscarPorId(String id) {
        Contrato contrato = contratoRepository.buscarPorId(id);

        if (contrato == null) {
            throw new RuntimeException("Contrato no encontrado");
        }

        return contrato;
    }

    public Contrato crear(Contrato contrato) {
        validarContrato(contrato);

        if (contrato.getEstado() == null || contrato.getEstado().isBlank()) {
            contrato.setEstado("ACTIVO");
        }

        return contratoRepository.guardar(contrato);
    }

    public Contrato actualizar(String id, Contrato contrato) {
        validarContrato(contrato);
        return contratoRepository.actualizar(id, contrato);
    }

    public void eliminar(String id) {
        contratoRepository.eliminar(id);
    }

    public Contrato[] listarPorEstado(String estado) {
        actualizarEstadosAutomaticos();
        return convertirAArreglo(contratoRepository.listarPorEstado(estado));
    }

    public Contrato[] listarPorCliente(String idCliente) {
        actualizarEstadosAutomaticos();
        return convertirAArreglo(contratoRepository.listarPorCliente(idCliente));
    }

    public Contrato[] listarProximosAVencer(int dias) {
        actualizarEstadosAutomaticos();
        return convertirAArreglo(contratoRepository.listarProximosAVencer(dias));
    }

    public Contrato[] listarVencidos() {
        actualizarEstadosAutomaticos();
        return convertirAArreglo(contratoRepository.listarVencidos());
    }

    public Contrato cambiarEstado(String id, String nuevoEstado) {
        Contrato contrato = buscarPorId(id);
        contrato.setEstado(nuevoEstado);
        return contratoRepository.actualizar(id, contrato);
    }

    private void validarContrato(Contrato contrato) {
        if (contrato.getId() == null || contrato.getId().isBlank()) {
            throw new RuntimeException("El ID del contrato es obligatorio");
        }

        if (contrato.getCodigoInmueble() == null || contrato.getCodigoInmueble().isBlank()) {
            throw new RuntimeException("El código del inmueble es obligatorio");
        }

        if (contrato.getIdCliente() == null || contrato.getIdCliente().isBlank()) {
            throw new RuntimeException("El ID del cliente es obligatorio");
        }

        if (contrato.getIdAsesor() == null || contrato.getIdAsesor().isBlank()) {
            throw new RuntimeException("El ID del asesor es obligatorio");
        }

        if (contrato.getTipoContrato() == null || contrato.getTipoContrato().isBlank()) {
            throw new RuntimeException("El tipo de contrato es obligatorio");
        }

        if (contrato.getFechaInicio() == null) {
            throw new RuntimeException("La fecha de inicio es obligatoria");
        }

        if (contrato.getFechaFin() == null) {
            throw new RuntimeException("La fecha de fin es obligatoria");
        }

        if (contrato.getFechaFin().isBefore(contrato.getFechaInicio())) {
            throw new RuntimeException("La fecha de fin no puede ser anterior a la fecha de inicio");
        }

        if (contrato.getValor() < 0) {
            throw new RuntimeException("El valor del contrato no puede ser negativo");
        }
    }

    public void actualizarEstadosAutomaticos() {
        LinkedSimpleList<Contrato> contratos = contratoRepository.listar();

        for (Contrato contrato : contratos) {
            String estadoOriginal = contrato.getEstado();

            if (contrato.getEstado() == null) {
                contrato.setEstado("ACTIVO");
            }

            if (contrato.estaVencido()
                    && !contrato.getEstado().equalsIgnoreCase("CANCELADO")
                    && !contrato.getEstado().equalsIgnoreCase("FINALIZADO")) {
                contrato.setEstado("VENCIDO");
            } else if (contrato.estaProximoAVencer(30)
                    && contrato.getEstado().equalsIgnoreCase("ACTIVO")) {
                contrato.setEstado("PROXIMO_A_VENCER");
            }

            if (estadoOriginal == null || !estadoOriginal.equalsIgnoreCase(contrato.getEstado())) {
                contratoRepository.actualizar(contrato.getId(), contrato);
            }
        }
    }

    private Contrato[] convertirAArreglo(LinkedSimpleList<Contrato> contratos) {
        Contrato[] resultado = new Contrato[contratos.getSize()];
        int indice = 0;

        for (Contrato contrato : contratos) {
            resultado[indice] = contrato;
            indice++;
        }

        return resultado;
    }
}
